stdlib/json.gr

```diff
@@ -0,0 +1,2103 @@
+ * characters for illustrative purposes.
+ *
+ * `CompactArrayEntries`
+ * ```
+ * []
+ * ```
+ *
+ * ```
+ * [1]
+ * ```
+ *
+ * ```
+ * [1,2,3]
+ * ```
+ *
+ * `SpacedArrayEntries`
+ * ```
+ * [ ]
+ * ```
+ *
+ * ```
+ * [1]
+ * ```
+ *
+ * ```
+ * [1,2,3]
+ * ```
+ *
+ * `OneArrayEntryPerLine`
+ * ```
+ * []
+ * ```
+ *
+ * ```
+ * [↵
+ * ··1↵
+ * ]
+ * ```
+ *
+ * ```
+ * [↵
+ * ··1,↵
+ * ··2,↵
+ * ··3↵
+ * ]
+ * ```
+ */
+provide enum ArrayFormat {
+  CompactArrayEntries,
+  SpacedArrayEntries,
+  OneArrayEntryPerLine,
+}
+
+/**
+ * Controls how objects are printed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * characters for illustrative purposes.
+ *
+ * `CompactObjectEntries`
+ * ```
+ * {}
+ * ```
+ *
+ * ```
+ * {"a":1}
+ * ```
+ *
+ * ```
+ * {"a":1,"b":2,"c":3}
+ * ```
+ *
+ * `SpacedObjectEntries`
+ * ```
+ * { }
+ * ```
+ *
+ * ```
+ * {"a": 1}
+ * ```
+ *
+ * ```
+ * {"a": 1, "b": 2, "c": 3}
+ * ```
+ *
+ * `OneObjectEntryPerLine`
+ * ```
+ * {}
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1↵
+ * }
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1,↵
+ * ··"b":·2,↵
+ * ··"c":·3↵
+ * }
+ * ```
+ */
+provide enum ObjectFormat {
+  CompactObjectEntries,
+  SpacedObjectEntries,
+  OneObjectEntryPerLine,
+}
+
+/**
+ * Controls line ending type in custom formatting.
+ */
+provide enum LineEnding {
+  NoLineEnding,
+  LineFeed,
+  CarriageReturnLineFeed,
+  CarriageReturn,
+}
+
+/*
+ * Allows fine-grained control of formatting in JSON printing.
+ */
+record FormattingSettings {
+  indentation: IndentationFormat,
+  arrayFormat: ArrayFormat,
+  objectFormat: ObjectFormat,
+  lineEnding: LineEnding,
+  finishWithNewLine: Bool,
+  escapeAllControlPoints: Bool,
+  escapeHTMLUnsafeSequences: Bool,
+  escapeNonASCII: Bool,
+}
+
+/**
+ * Allows control of formatting in JSON printing.
+ */
+provide enum FormattingChoices {
+  /**
+   * Recommended human readable formatting.
+   *
+   * Escapes all control points for the sake of clarity, but prints unicode
+   * codepoints directly so the result needs to be treated as proper unicode and
+   * is not safe to be transported in ASCII encoding.
+   *
+   * Roughly Equivalent to:
+   * ```
+   * Custom{
+   *  indentation: IndentWithSpaces(2),
+   *  arrayFormat: OneArrayEntryPerLine,
+   *  objectFormat: OneObjectEntryPerLine,
+   *  lineEnding: LineFeed,
+   *  finishWithNewLine: true,
+   *  escapeAllControlPoints: true,
+   *  escapeHTMLUnsafeSequences: false,
+   *  escapeNonASCII: false,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```
+   * {↵
+   * ··"currency":·"€",↵
+   * ··"price":·99.9,↵
+   * ··"currencyDescription":·"EURO\u007f",↵
+   * ··"script·unembeddable":·"You·cannot·paste·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·</·with·unescaped·forward·slash"↵
```

Comment:
What is happening here?

Reply:
Was just an example, I removed that line though as I think its a little misleading.